File: src/237.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "list.h"

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
void deleteNode(struct ListNode* node) {
    if ((node == NULL) || (node->next == NULL)) {
        return;
    }

    struct ListNode* p = node->next;
    *node = *node->next;
    free(p);
}

int main(int argc, char** argv) {
#define MAX_SIZE (1000)
    struct testCaseType {
        int node[MAX_SIZE];
        int nodeSize;
        int deleteVal;
    } testCase[] = {{{4, 5, 1, 9}, 4, 5}, {{4, 5, 1, 9}, 4, 1}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: head = [4,5,1,9], node = 5
     *  Output: [4,1,9]
     *
     *  Input: head = [4,5,1,9], node = 1
     *  Output: [4,5,9]
     */

    struct ListNode* pNode = NULL;
    struct ListNode* pHead = NULL;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        for (j = 0; j < testCase[i].nodeSize; ++j) {
            pHead = addValueToEndListNode(pHead, testCase[i].node[j]);
        }
        printf("Input: head = [");
        printList(pHead);
        printf("], node = %d\n", testCase[i].deleteVal);

        pNode = pHead;
        while (pNode != NULL) {
            if (pNode->val == testCase[i].deleteVal) {
                break;
            }
            pNode = pNode->next;
        }

        deleteNode(pNode);
        printf("Output: [");
        printList(pHead);
        printf("]\n");

        printf("\n");

        freeListNode(pHead);
        pHead = NULL;
    }

    return EXIT_SUCCESS;
}

File: src/237.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop

from list import ListNode, ListNodeCommon


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def deleteNode(self, node):
        """
        :type node: ListNode
        :rtype: void Do not return anything, modify node in-place instead.
        """
        node.val = node.next.val
        node.next = node.next.next


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for head, node in zip([[4, 5, 1, 9], [4, 5, 1, 9]], [5, 1]):
            # /* Example
            #  *  Input: head = [4,5,1,9], node = 5
            #  *  Output: [4,1,9]
            #  *
            #  *  Input: head = [4,5,1,9], node = 1
            #  *  Output: [4,5,9]
            #  */
            logging.info("Input: head = %s, node = %s", head, node)

            pHead = ListNode.buildList(head)
            pNode = pHead
            while pNode != None:
                if pNode.val == node:
                    break
                pNode = pNode.next
            pSolution.deleteNode(pNode)
            logging.info("Output: %s", pHead)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
